drivers/flash/soc_flash_numaker.c

```diff
@@ -0,0 +1,290 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ *  Copyright (c) 2023 Nuvoton Technology Corporation.
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_fmc
+
+#include <string.h>
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+#include "flash_priv.h"
+#include <NuMicro.h>
+
+LOG_MODULE_REGISTER(flash_numaker, CONFIG_FLASH_LOG_LEVEL);
+
+#define SOC_NV_FLASH_NODE DT_INST(0, soc_nv_flash)
+
+extern int32_t FMC_EraseBank(uint32_t u32_bank_addr);
+
+struct flash_numaker_data {
+	FMC_T *fmc;
+	struct k_sem write_lock;
+	uint32_t flash_block_base;
+};
+
+static const struct flash_parameters flash_numaker_parameters = {
+#if DT_NODE_HAS_PROP(SOC_NV_FLASH_NODE, write_block_size)
+	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
+#else
+	.write_block_size = 0x04,
+#endif
+	.erase_value = 0xff,
+};
+
+/* Validate offset and length */
+static bool flash_numaker_is_range_valid(off_t offset, size_t len)
+{
+	/* check for min value */
+	if ((offset < 0) || (len < 1)) {
+		return false;
+	}
+
+	/* check for max value */
+	if ((offset + len) > (FMC_APROM_END - FMC_APROM_BASE)) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Erase a flash memory area.
+ *
+ * param dev       Device struct
+ * param offset    The address's offset
+ * param len       The size of the buffer
+ * return 0       on success
+ * return -EINVAL erroneous code
+ */
+
+static int flash_numaker_erase(const struct device *dev, off_t offset, size_t len)
+{
+	struct flash_numaker_data *dev_data = dev->data;
+	uint32_t rc = 0;
+	unsigned int key;
+	int page_nums = (len / FMC_FLASH_PAGE_SIZE);
+	uint32_t addr = dev_data->flash_block_base + offset;
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
+		return 0;
+	}
+
+	/* Validate range */
+	if (!flash_numaker_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* check alignment and erase only by pages */
+	if (((addr % FMC_FLASH_PAGE_SIZE) != 0) || ((len % FMC_FLASH_PAGE_SIZE) != 0)) {
+		return -EINVAL;
+	}
+
+	/* take semaphore */
+	if (k_sem_take(&dev_data->write_lock, K_NO_WAIT)) {
+		return -EACCES;
+	}
+
+	SYS_UnlockReg();
+	key = irq_lock();
+	while (page_nums) {
+		if (((len >= FMC_BANK_SIZE)) && ((addr % FMC_BANK_SIZE) == 0)) {
+			if (FMC_EraseBank(addr)) {
+				LOG_ERR("Erase flash bank failed or erase time-out");
+				rc = -EIO;
+				goto move_exit;
+			}
+			page_nums -= (FMC_BANK_SIZE / FMC_FLASH_PAGE_SIZE);
+			addr += FMC_BANK_SIZE;
+		} else {
+			/* erase page */
+			if (FMC_Erase(addr)) {
+				LOG_ERR("Erase flash page failed or erase time-out");
+				rc = -EIO;
+				goto move_exit;
+			}
+			page_nums--;
+			addr += FMC_FLASH_PAGE_SIZE;
+		}
+	}
+
+move_exit:
+	SYS_LockReg();
+	irq_unlock(key);
```

Comment:
is the order of this correct/important? Would have expected a symmetry between SYS_... and irq_...

Reply:
Yes, the order is important and expected the pair of SYS_UnlockReg v.s SYS_LockReg, irq_lock v.s irq_unlock.